File: prime_sieve/sieve.cs
using System;
using System.Collections.Generic;

class Program
{
    static List<int> SieveOfEratosthenes(int n)
    {
        if (n < 2)
        {
            return new List<int>();
        }
        
        // Create boolean array
        bool[] primes = new bool[n + 1];
        for (int i = 2; i <= n; i++)
        {
            primes[i] = true;
        }
        
        // Sieve algorithm
        for (int i = 2; i * i <= n; i++)
        {
            if (primes[i])
            {
                for (int j = i * i; j <= n; j += i)
                {
                    primes[j] = false;
                }
            }
        }
        
        // Collect prime numbers
        List<int> result = new List<int>();
        for (int i = 2; i <= n; i++)
        {
            if (primes[i])
            {
                result.Add(i);
            }
        }
        
        return result;
    }
    
    static void Main()
    {
        int limit = 10000000;  // 10 million
        List<int> primes = SieveOfEratosthenes(limit);
        
        Console.WriteLine($"Found {primes.Count} primes up to {limit}");
        Console.Write("First 10 primes: ");
        for (int i = 0; i < 10 && i < primes.Count; i++)
        {
            Console.Write($"{primes[i]} ");
        }
        Console.WriteLine();
        
        Console.Write("Last 10 primes: ");
        for (int i = primes.Count - 10; i < primes.Count; i++)
        {
            Console.Write($"{primes[i]} ");
        }
        Console.WriteLine();
    }
}

File: prime_sieve/sieve.rs
use std::collections::VecDeque;

fn sieve_of_eratosthenes(n: usize) -> Vec<i32> {
    if n < 2 {
        return vec![];
    }
    
    // Create boolean vector
    let mut primes = vec![true; n + 1];
    primes[0] = false;
    primes[1] = false;
    
    // Sieve algorithm
    for i in 2..=((n as f64).sqrt() as usize) {
        if primes[i] {
            for j in (i * i..=n).step_by(i) {
                primes[j] = false;
            }
        }
    }
    
    // Collect prime numbers
    let mut result = Vec::new();
    for i in 2..=n {
        if primes[i] {
            result.push(i as i32);
        }
    }
    
    result
}

fn main() {
    let limit = 10_000_000;  // 10 million
    let primes = sieve_of_eratosthenes(limit);
    
    println!("Found {} primes up to {}", primes.len(), limit);
    print!("First 10 primes: ");
    for i in 0..10.min(primes.len()) {
        print!("{} ", primes[i]);
    }
    println!();
    
    print!("Last 10 primes: ");
    for i in (primes.len().saturating_sub(10))..primes.len() {
        print!("{} ", primes[i]);
    }
    println!();
}
